### PR TITLE
Introduced the Tenant Group for Alfred Throttler

### PR DESCRIPTION
**Pull request by tianzhu007**
We are going to add throttling to some specific tenant ids, such as throttle PP tenants more and protect other tenants like Hexagon. We would like to get the throttling limits like this:

<html>
<body>
<span><span class="ui-provider fe cae bfw caf cag cah cai caj cak cal cam can cao cap caq car cas cat cau cav caw cax cay caz cba cbb cbc cbd cbe cbf cbg cbh cbi cbj cbk" dir="ltr">

  | create Doc | delete Doc | ....
-- | -- | -- | --
tenant A, tenant B, .... | x | y | ...
tenant C | a | b | ...
... | ... | ... | ...

</span></span>
</body>
</html>

Right now, we plan to introduce tenant groups consists different tenant ids, and their throttling limits in the config files to implement this feature in the Alfred endpoints.

Also, this PR adjust the throttling order to check throttling before validating tokens.

### CODE REVIEW

File: server/routerlicious/packages/lambdas/package.json [hunk 1]
```diff
@@ -119,8 +119,14 @@
 				"forwardCompat": false,
 				"backCompat": false
 			},
-			"RemovedClassDeclaration_ForemanLambda": {"forwardCompat": false, "backCompat": false},
-			"RemovedClassDeclaration_ForemanLambdaFactory": {"forwardCompat": false, "backCompat": false}
+			"RemovedClassDeclaration_ForemanLambda": {
```
Comment (tianzhu007) [inline server/routerlicious/packages/lambdas/package.json:122]: @znewton Need a review here. :)

File: server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts [hunk 2]
```diff
@@ -264,89 +260,163 @@ export class AlfredResourcesFactory implements core.IResourcesFactory<AlfredReso
...
+		restClusterThrottlers.set(
+			Constants.submitSignalThrottleIdPrefix,
+			socketSubmitSignalThrottler,
 		);
 
-		const socketConnectionThrottleConfigPerCluster = utils.getThrottleConfig(
-			config.get("alfred:throttling:socketConnectionsPerCluster"),
+		// Tenant Group 1 Rest API Throttlers
+		const restApiCreateDocThrottleConfigGroup1 = utils.getThrottleConfig(
+			config.get("alfred:throttling:generalCluster:createDoc"),
```
Comment (znewton) [inline server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts:362]: All of the configs from here and below should be using `tenantGroup1` instead of `generalCluster`, I think

**Review by znewton — COMMENTED** [review]
quick look

File: server/routerlicious/packages/lambdas/package.json (hunk 1 above)
Comment (znewton) [reply server/routerlicious/packages/lambdas/package.json:122]: Type validation configs look fine

File: server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts [hunk 3]
```diff
@@ -264,89 +260,163 @@ export class AlfredResourcesFactory implements core.IResourcesFactory<AlfredReso
 			);
 		};
 
+		const tenantThrottlersMap = new Map<string, string>();
+		const tenantGroup1 = config.get("alfred:throttlingGroup:tenantGroup1") as string[];
+		tenantGroup1.forEach((t) => tenantThrottlersMap.set(t, "tenantGroup1"));
```
Comment (znewton) [inline server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts:265]: This is a good first pass. I think we should do a follow-up PR to make this more flexible so that we can add more tenantGroups to the configs as necessary, without having to add more code each time. Not 100% necessary on this first PR though.

File: server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts [hunk 4]
```diff
@@ -78,6 +71,27 @@ export function create(
 		throttleIdSuffix: Constants.alfredRestThrottleIdSuffix,
 	};
 
+	// Todo: delete it and use the function in throttlerMiddleware.ts
+	const tenantThrottle = (throttleApi: string) => {
```
Comment (znewton) [inline server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts:75]: R11s-base can use throttlerMiddleware from services-utills right now. Historian is the only place you need to duplicate this.

File: server/routerlicious/packages/routerlicious-base/src/test/alfred/io.spec.ts [hunk 5]
```diff
@@ -583,18 +610,44 @@ Submitted Messages: ${JSON.stringify(messages, undefined, 2)}`,
...
+					const testGroup1SubmitOpThrottler = new TestThrottler(throttleLimitConnectDoc);
 
-					const testSubmitSignalThrottler = new Throttler(
+					const throttlerHelper = new ThrottlerHelper(testThrottleAndUsageStorageManager);
+					const testSubmitSignalThrottlerPerCluster = new Throttler(
 						throttlerHelper,
 						minThrottleCheckInterval,
 					);
 
+					tenantThrottlers.set("socketConnections", testConnectionThrottlerPerTenant);
```
Comment (znewton) [inline server/routerlicious/packages/routerlicious-base/src/test/alfred/io.spec.ts:635]: Should these be using the prefixes defined in Constants?

File: server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts (hunk 4 above)
Comment (tianzhu007) [reply server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts]: Thanks. Resolved. :)

File: server/routerlicious/packages/services-utils/src/throttlerMiddleware.ts [hunk 6]
```diff
@@ -56,6 +57,34 @@ function noopMiddleware(req: Request, res: Response, next: NextFunction) {
...
+			throttleIdPrefix: tenantGroup ? `${tenantId}_${tenantGroup}` : tenantId,
+			throttleIdSuffix: throttleApi,
+		};
+		const throttler = tenantGroup
+			? throttlersMap.get(tenantGroup)?.get(throttleApi)
+			: throttlersMap.get("generalTenant")?.get(throttleApi);
+		if (throttler) {
+			return throttle(throttler, undefined, throttleOptions)(req, rest, next);
+		}
+		next();
```
Comment (tianzhu007) [inline server/routerlicious/packages/services-utils/src/throttlerMiddleware.ts:84]: @znewton  Need a review for this method, and the next(). Thanks. :)

**Review by znewton — APPROVED** [review]

File: server/routerlicious/packages/lambdas/src/alfred/index.ts [hunk 7]
```diff
@@ -269,17 +267,25 @@ export function configureWebSocketServices(
 		async function connectDocument(message: IConnect): Promise<IConnectedClient> {
 			const startTime = Date.now();
 			const throttleErrorPerCluster = checkThrottleAndUsage(
-				connectThrottlerPerCluster,
+				throttlersMap?.get("generalCluster")?.get("socketConnections"),
```
Comment (sharmakhushboo) [inline server/routerlicious/packages/lambdas/src/alfred/index.ts:270]: just confirming - Does checkThrottleAndUsage() gracefully handle null or undefined throttler?
It might be a good idea to test with empty throttler or missing throttling in map or empty throttlerMap in config ?

File: server/routerlicious/packages/lambdas/src/alfred/index.ts [hunk 8]
```diff
@@ -269,17 +267,25 @@ export function configureWebSocketServices(
...
 				getSocketConnectThrottleId("connectDoc"),
 				message.tenantId,
 				logger,
 			);
 			if (throttleErrorPerCluster) {
 				return Promise.reject(throttleErrorPerCluster);
 			}
+
```
Comment (sharmakhushboo) [inline server/routerlicious/packages/lambdas/src/alfred/index.ts:278]: This code is getting complicated with each iteration. :D It would be great if you could add some code comments to make it readable for other developers?

For example here, you can add something like:

If tenant is found in any tenantGroup list, apply tenantGroup specific throttling limits, else apply general per tenant throttling limits.

Feel free to edit or add more details ^

File: server/routerlicious/kubernetes/routerlicious/values.yaml [hunk 9]
```diff
@@ -20,23 +20,37 @@ alfred:
...
-        maxPerMs: 1000000
-        maxBurst: 1000000
-        minCooldownIntervalInMs: 1000000
-        minThrottleIntervalInMs: 1000000
-        maxInMemoryCacheSize: 1000
-        maxInMemoryCacheAgeInMs: 60000
-        enableEnhancedTelemetry: false
-    restCallsPerCluster:
-        createDoc: 
+    tenantGroup1:
```
Comment (sharmakhushboo) [inline server/routerlicious/kubernetes/routerlicious/values.yaml:23]: This naming is a little confusing. It would be good to mention somewhere that tenantGroup limits are still per Tenant and not per group.

File: server/routerlicious/packages/lambdas/src/alfred/index.ts [hunk 10]
```diff
@@ -269,17 +267,25 @@ export function configureWebSocketServices(
...
-				getSocketConnectThrottleId(message.tenantId),
+				tenantGroup
+					? throttlersMap?.get(tenantGroup)?.get("socketConnections")
+					: throttlersMap?.get("generalTenant")?.get("socketConnections"),
+				tenantGroup
+					? `${getSocketConnectThrottleId(message.tenantId)}_${tenantGroup}`
+					: getSocketConnectThrottleId(message.tenantId),
 				message.tenantId,
 				logger,
 			);
```
Comment (sharmakhushboo) [inline server/routerlicious/packages/lambdas/src/alfred/index.ts:291]: We also have clients per socket connection check on line number 419. Could you also move that check after perTenant throttling? or you can do it as part of your next PR when you move token validation after throttling check for other apis.

`429,
					"Too Many Clients Connected to Document"`

File: server/routerlicious/packages/lambdas/src/alfred/index.ts [hunk 11]
```diff
@@ -269,17 +267,25 @@ export function configureWebSocketServices(
...
 				return Promise.reject(throttleErrorPerCluster);
 			}
+
+			const tenantGroup: string | undefined = message.tenantId
+				? tenantThrottlersMap?.get(message.tenantId) ?? undefined
+				: undefined;
 			const throttleErrorPerTenant = checkThrottleAndUsage(
-				connectThrottlerPerTenant,
-				getSocketConnectThrottleId(message.tenantId),
+				tenantGroup
```
Comment (sharmakhushboo) [inline server/routerlicious/packages/lambdas/src/alfred/index.ts:283]: Can we change this
`const throttleErrorPerTenant = checkThrottleAndUsage(
				tenantGroup
					? throttlersMap?.get(tenantGroup)?.get("socketConnections")
					: throttlersMap?.get("generalTenant")?.get("socketConnections"),
				tenantGroup
					? `${getSocketConnectThrottleId(message.tenantId)}_${tenantGroup}`
					: getSocketConnectThrottleId(message.tenantId),
				message.tenantId,
				logger,
			);`

to
`const connectThrottlerPerTenant = tenantGroup
					? throttlersMap?.get(tenantGroup)?.get("socketConnections")
					: throttlersMap?.get("generalTenant")?.get("socketConnections");

const connectThrottlerPerTenantId = tenantGroup
					? `${getSocketConnectThrottleId(message.tenantId)}_${tenantGroup}`
					: getSocketConnectThrottleId(message.tenantId);

const throttleErrorPerTenant = checkThrottleAndUsage(
				connectThrottlerPerTenant,
				connectThrottlerPerTenantId,
				message.tenantId,
				logger,
			);`

basically moving the null checks(?:) out of the checkThrottleAndUsage(). This will make the code more readable.

Comment (sharmakhushboo) [reply server/routerlicious/packages/lambdas/src/alfred/index.ts:283]: Looks like github didn't format all the code. let me know if my comment seems confusing.

File: server/routerlicious/kubernetes/routerlicious/values.yaml [hunk 12]
```diff
@@ -20,23 +20,37 @@ alfred:
...
+        socketConnections: disabled
+        submitOps: disabled
+    generalTenant:
+        generalRestCall: disabled
+        createDoc: disabled
+        getDeltas: disabled
+        getSession: disabled
+        socketConnections: disabled
+        submitOps: disabled
+    generalCluster:
```
Comment (sharmakhushboo) [inline server/routerlicious/kubernetes/routerlicious/values.yaml:43]: Let's call this perCluster. We are not splitting this to perTenant, so generalCluster does not make sense here

File: server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts [hunk 13]
```diff
@@ -264,89 +260,163 @@ export class AlfredResourcesFactory implements core.IResourcesFactory<AlfredReso
...
+
```
Comment (sharmakhushboo) [inline server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts:266]: I agree with Zach here. We will need to make this design more flexible as right now tenantGroup1 is hardcoded which means we can only have 1 group and also, we cannot rename the group to make it more readable. @tianzhu007 could you please add a backlog workitem for this? Thank you :)

File: server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts [hunk 14]
```diff
@@ -264,89 +260,163 @@ export class AlfredResourcesFactory implements core.IResourcesFactory<AlfredReso
...
+		const socketConnectionThrottleConfigPerClusterGroup1 = utils.getThrottleConfig(
+			config.get("alfred:throttling:tenantGroup1:socketConnections"),
+		);
+		const socketConnectClusterThrottlerGroup1 = configureThrottler(
+			socketConnectionThrottleConfigPerClusterGroup1,
+		);
+		const submitOpThrottleConfigGroup1 = utils.getThrottleConfig(
+			config.get("alfred:throttling:tenantGroup1:submitOps"),
+		);
+		const socketSubmitOpThrottlerGroup1 = configureThrottler(submitOpThrottleConfigGroup1);
```
Comment (sharmakhushboo) [inline server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts:388]: we can just call this submitOpThrottlerGroup1. same for signals

File: server/routerlicious/packages/routerlicious-base/src/utils/constants.ts [hunk 15]
```diff
@@ -8,5 +8,11 @@ export const Constants = Object.freeze({
 	createDocThrottleIdPrefix: "createDoc",
 	getDeltasThrottleIdPrefix: "getDeltas",
 	getSessionThrottleIdPrefix: "getSession",
+	socketConnectionsThrottleIdPrefix: "socketConnections",
+	submitOpsThrottleIdPrefix: "submitOps",
+	submitSignalThrottleIdPrefix: "submitSignal",
 	generalRestCallThrottleIdPrefix: "generalRestCall",
+	throttleGeneralTenant: "generalTenant",
+	throttleGeneralCluster: "generalCluster",
+	throttleTenantGroup1: "tenantGroup1",
```
Comment (sharmakhushboo) [inline server/routerlicious/packages/routerlicious-base/src/utils/constants.ts:17]: The last three are throttler names and Not prefixes, right?
 
If yes, then maybe we should name them differently. constant variable name could be something like:
generalTenantThrottler: "generalTenant",
generalClusterThrottler: "generalCluster",
tenantGroup1Throttler: "tenantGroup1",


I do feel this list is growing and we will need to refactor a lot of this when we have some more bandwidth :(

File: server/routerlicious/kubernetes/routerlicious/values.yaml (hunk 9 above)
Comment (sharmakhushboo) [reply server/routerlicious/kubernetes/routerlicious/values.yaml:23]: Another thought: since we can only have 1 tenantGroup right now (because this isn't extensible right now), should we call this exceptionTenants, or something similar?

Reason -> i feel saying tenantGroup1 without having flexibility of creating another tenantGroup is misleading.

File: server/routerlicious/packages/lambdas/src/alfred/index.ts (hunk 11 above)
Comment (tianzhu007) [reply server/routerlicious/packages/lambdas/src/alfred/index.ts]: Done. :)

File: server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts (hunk 14 above)
Comment (tianzhu007) [reply server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts]: Sounds good! Done. :)

File: server/routerlicious/packages/routerlicious-base/src/utils/constants.ts (hunk 15 above)
Comment (tianzhu007) [reply server/routerlicious/packages/routerlicious-base/src/utils/constants.ts]: Sounds good! Done. :)

File: server/routerlicious/kubernetes/routerlicious/values.yaml (hunk 12 above)
Comment (tianzhu007) [reply server/routerlicious/kubernetes/routerlicious/values.yaml]: Sounds good! Done. :)

File: server/routerlicious/packages/lambdas/src/alfred/index.ts (hunk 10 above)
Comment (tianzhu007) [reply server/routerlicious/packages/lambdas/src/alfred/index.ts:293]: I just got a check, and the token validation happened after throttling the check here. :)

File: server/routerlicious/packages/lambdas/src/alfred/index.ts (hunk 8 above)
Comment (tianzhu007) [reply server/routerlicious/packages/lambdas/src/alfred/index.ts:278]: Thanks. Done. :)

File: server/routerlicious/packages/lambdas/src/alfred/index.ts (hunk 7 above)
Comment (tianzhu007) [reply server/routerlicious/packages/lambdas/src/alfred/index.ts]: Yes, the checkThrottleAndUsage() will directly return for handling null or undefined throttler. 
Current throttlerMap is hard-coded, so it is less chance to let it be null or empty.

File: server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts (hunk 13 above)
Comment (tianzhu007) [reply server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts:266]: Sounds good! Made a work item here
https://msazure.visualstudio.com/One/_workitems/edit/21283047

File: server/routerlicious/kubernetes/routerlicious/values.yaml (hunk 9 above)
Comment (tianzhu007) [reply server/routerlicious/kubernetes/routerlicious/values.yaml]: Change the name to tenantsGroup1PerTenant.
Also, there is a possibility to scale out tenants group, such as tenantsGroup2PerTenant, and so on.

File: server/routerlicious/packages/lambdas/src/alfred/index.ts [hunk 16]
```diff
@@ -801,7 +840,7 @@ export function configureWebSocketServices(
 						clientId,
 					};
 					const throttleError = checkThrottleAndUsage(
-						submitSignalThrottler,
+						throttlersMap?.get("perCluster")?.get("submitSignal"),
 						getSubmitSignalThrottleId(clientId, room.tenantId),
```
Comment (sharmakhushboo) [inline server/routerlicious/packages/lambdas/src/alfred/index.ts:844]: I think this should be "submitSignals". perCluster throttleId should not be based on clientId+tenantId right?

File: server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/api.ts [hunk 17]
```diff
@@ -48,7 +48,9 @@ export function create(
 		throttleIdPrefix: (req) => getParam(req.params, "tenantId"),
 		throttleIdSuffix: Constants.alfredRestThrottleIdSuffix,
 	};
-	const generalTenantThrottler = tenantThrottlers.get(Constants.generalRestCallThrottleIdPrefix);
+	const perTenantThrottler = throttlersMap
+		.get("perTenant")
```
Comment (sharmakhushboo) [inline server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/api.ts:52]: this should be .get(Constants.perTenantThrottler)

File: server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/deltas.ts [hunk 18]
```diff
@@ -92,8 +93,8 @@ export function create(
 	router.get(
 		"/raw/:tenantId/:id",
 		validateRequestParams("tenantId", "id"),
+		throttle(perTenantThrottler, winston, tenantThrottleOptions),
```
Comment (sharmakhushboo) [inline server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/deltas.ts:96]: Can we not call tenantThrottle() middleware here?

File: server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts [hunk 19]
```diff
@@ -262,9 +262,9 @@ export function create(
 	router.post(
 		"/:tenantId/document/:id/revokeToken",
 		validateRequestParams("tenantId", "id"),
+		throttle(perTenantThrottler, winston, tenantThrottleOptions),
```
Comment (sharmakhushboo) [inline server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts:265]: Why not tenantThrottler() middleware?

Also, let's also add perCluster throttle for this api.

File: server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts [hunk 20]
```diff
@@ -299,89 +295,167 @@ export class AlfredResourcesFactory implements core.IResourcesFactory<AlfredReso
 			);
 		};
 
+		const tenantThrottlersMap = new Map<string, string>();
+		const tenantsGroup1PerTenant = config.get("shared:tenantsGroup1PerTenant") as string[];
+		tenantsGroup1PerTenant.forEach((t) => tenantThrottlersMap.set(t, "tenantsGroup1PerTenant"));
+
+		const throttlersMap = new Map<string, Map<string, core.IThrottler>>();
```
Comment (sharmakhushboo) [inline server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts:302]: Can we please move all this code to a separate method in a separate .ts file or utils file?

The util methods can be 
1. getTenantGroupMap() and,
2. getThrottlerMap().

File: server/routerlicious/kubernetes/routerlicious/values.yaml (hunk 9 above)
Comment (sharmakhushboo) [reply server/routerlicious/kubernetes/routerlicious/values.yaml]: Correct but to do that, we will also have to update runnerFactory as well. So, it won't just be a config update, we will also have to change the code.

so essentially right now, with this pr code change, we can only support 1 tenant Group

File: server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts [hunk 21]
```diff
@@ -299,89 +295,167 @@ export class AlfredResourcesFactory implements core.IResourcesFactory<AlfredReso
...
+		clusterThrottlers.set(Constants.getSessionThrottleIdPrefix, restGetSessionThrottler);
+		clusterThrottlers.set(
+			Constants.socketConnectionsThrottleIdPrefix,
+			socketConnectClusterThrottler,
+		);
+		clusterThrottlers.set(Constants.submitOpsThrottleIdPrefix, socketSubmitOpThrottler);
+		clusterThrottlers.set(Constants.submitSignalThrottleIdPrefix, socketSubmitSignalThrottler);
+
+		// Tenant Group 1 Rest API Throttlers
+		const restApiCreateDocThrottleConfigGroup1 = utils.getThrottleConfig(
```
Comment (sharmakhushboo) [inline server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts:393]: (*non-blocking*)
I think we can remove rest prefix and add tenant to the name instead

> CreateDocThrottleConfigTenantGroup1

File: server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts [hunk 22]
```diff
@@ -299,89 +295,167 @@ export class AlfredResourcesFactory implements core.IResourcesFactory<AlfredReso
...
+			config.get("alfred:throttling:perTenant:socketConnections"),
+		);
+		const socketConnectTenantThrottler = configureThrottler(
+			socketConnectionThrottleConfigPerTenant,
+		);
+		const submitOpThrottleConfigPerTenant = utils.getThrottleConfig(
+			config.get("alfred:throttling:perTenant:submitOps"),
+		);
+		const socketSubmitOpTenantThrottler = configureThrottler(submitOpThrottleConfigPerTenant);
 		const restTenantThrottlers = new Map<string, core.IThrottler>();
```
Comment (sharmakhushboo) [inline server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts:336]: just call it tenantThrottlers

**Review by sharmakhushboo — CHANGES_REQUESTED** [review]

File: server/routerlicious/packages/lambdas/src/alfred/index.ts (hunk 10 above)
Comment (tianzhu007) [reply server/routerlicious/packages/lambdas/src/alfred/index.ts:294]: Got it. I had a try, the clients.length would depend on the call at line 398, which is hard to move that code up.

File: server/routerlicious/kubernetes/routerlicious/templates/fluid-configmap.yaml [hunk 23]
```diff
@@ -73,21 +73,23 @@ data:
...
-                    "getDeltas": {{ toJson .Values.alfred.throttling.restCallsPerTenant.getDeltas }},
-                    "getSession": {{ toJson .Values.alfred.throttling.restCallsPerTenant.getSession }}}
+                "tenantsGroup1PerTenant" : {{ toJson .Values.alfred.throttling.tenantsGroup1PerTenant }},
+                "perTenant" : {
+                    "generalRestCall": {{ toJson .Values.alfred.throttling.perTenant.generalRestCall }},
+                    "createDoc": {{ toJson .Values.alfred.throttling.perTenant.createDoc }},
+                    "getDeltas": {{ toJson .Values.alfred.throttling.perTenant.getDeltas }},
+                    "getSession": {{ toJson .Values.alfred.throttling.perTenant.getSession }},
+                    "socketConnections": {{ toJson .Values.alfred.throttling.perTenant.socketConnections }},
+                    "submitOps": {{ toJson .Values.alfred.throttling.perTenant.submitOps }}
```
[the remark this answers is not in the document]
Comment (tianzhu007) [reply server/routerlicious/kubernetes/routerlicious/templates/fluid-configmap.yaml:83]: It is not required right now. :)

File: server/routerlicious/packages/lambdas/src/alfred/index.ts (hunk 16 above)
Comment (tianzhu007) [reply server/routerlicious/packages/lambdas/src/alfred/index.ts:844]: Got it. This should be the tenant level. :)

Comment (tianzhu007) [reply server/routerlicious/packages/lambdas/src/alfred/index.ts:844]: Changed to tenant level.

File: server/routerlicious/kubernetes/routerlicious/templates/fluid-configmap.yaml [hunk 24]
```diff
@@ -73,21 +73,23 @@ data:
             "sessionStickinessDurationMs": {{ .Values.alfred.sessionStickinessDurationMs }},
             "enableConnectionCountLogging": {{ .Values.alfred.enableConnectionCountLogging }},
             "throttling": {
-                "restCallsPerTenant": {
-                    "generalRestCall": {{ toJson .Values.alfred.throttling.restCallsPerTenant.generalRestCall }},
-                    "createDoc": {{ toJson .Values.alfred.throttling.restCallsPerTenant.createDoc }},
-                    "getDeltas": {{ toJson .Values.alfred.throttling.restCallsPerTenant.getDeltas }},
-                    "getSession": {{ toJson .Values.alfred.throttling.restCallsPerTenant.getSession }}}
+                "tenantsGroup1PerTenant" : {{ toJson .Values.alfred.throttling.tenantsGroup1PerTenant }},
```
[the remark this answers is not in the document]
Comment (tianzhu007) [reply server/routerlicious/kubernetes/routerlicious/templates/fluid-configmap.yaml]: Sounds good! Done.

File: server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts (hunk 22 above)
Comment (tianzhu007) [reply server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts]: Sounds good! Done. :)

File: server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/deltas.ts (hunk 18 above)
Comment (tianzhu007) [reply server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/deltas.ts]: Changed to perTenantGeneralThrottler, and this throttler is not enabled. :)

File: server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts (hunk 19 above)
Comment (tianzhu007) [reply server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts]: Changed to perTenantGeneralThrottler, and this throttler is not enabled. :)

Do we need to throttle this API? Maybe create a work item to enable the throttle for this API?

File: server/routerlicious/packages/lambdas/src/alfred/index.ts [hunk 25]
```diff
@@ -223,10 +223,8 @@ export function configureWebSocketServices(
 	isClientConnectivityCountingEnabled: boolean = false,
 	isSignalUsageCountingEnabled: boolean = false,
 	cache?: core.ICache,
-	connectThrottlerPerTenant?: core.IThrottler,
-	connectThrottlerPerCluster?: core.IThrottler,
-	submitOpThrottler?: core.IThrottler,
-	submitSignalThrottler?: core.IThrottler,
+	tenantThrottlersMap?: Map<string, string>,
```
[the remark this answers is not in the document]
Comment (tianzhu007) [reply server/routerlicious/packages/lambdas/src/alfred/index.ts]: Done. :)

File: server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts (hunk 21 above)
Comment (tianzhu007) [reply server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts]: Done. :)

File: server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/api.ts (hunk 17 above)
Comment (tianzhu007) [reply server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/api.ts]: Done. :)

File: server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts (hunk 20 above)
Comment (tianzhu007) [reply server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts:303]: I created the methods for throttlersMap.
I didn't create the tenant group map method, since it is straightforward.

Comment (tianzhu007) [reply server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts:303]: I created util methods for throttlers but doesn't do it for the tenant group, since it is straightforward.

File: server/routerlicious/kubernetes/routerlicious/values.yaml (hunk 9 above)
Comment (tianzhu007) [reply server/routerlicious/kubernetes/routerlicious/values.yaml]: Yes. We can keep it as it is right now. :) If we want to add more tenant groups, we will change the code.

Comment (tianzhu007) [reply server/routerlicious/kubernetes/routerlicious/values.yaml]: Reason: If we call it exceptionTenants right now, and add more tenant groups later, we would also change the code as well. :)
